[libc] Support %lc in printf

shubhe25p opened the pull request:
Add support for %lc in printf by calling internal wcrtomb function and relevant end-to-end sprintf test. Additionally, modified printf parser for recognizing length modifier and added two internal error codes. Also added a flag to disable wchar support on windows platform. Resolves #166598

vonosmas commented on libc/src/stdio/printf_core/char_converter.h:
```diff
@@ -33,7 +39,21 @@ LIBC_INLINE int convert_char(Writer<write_mode> *writer,
 
-  RET_IF_RESULT_NEGATIVE(writer->write(c));
+  if (to_conv.length_modifier == LengthModifier::l) {
+    wc = static_cast<wchar_t>(static_cast<unsigned int>(to_conv.conv_val_raw));
```
The argument type would be `wint_t` per standard, so you might need to handle WEOF and then static_cast.

You mentioned `BigInt` problems in the review thread, could you elaborate how are these relevant?

shubhe25p replied:
Thanks this makes sense, what I was referring to was that naively casting from `conv_val_raw` (BigInt) to `wchar_t` leads to an error in `BigInt to()`, as the casting has a guard clause that essentially checks:

```
struct BigInt {
...
template <typename T>
  LIBC_INLINE constexpr cpp::enable_if_t<
      cpp::is_integral_v<T> && !cpp::is_same_v<T, bool>, T>
  to() const { 
       ... }
```
  
`  is_integral_v ` will return false if the type is not one of:
  
``` 
char, signed char, unsigned char, short, unsigned short, int,
      unsigned int, long, unsigned long, long long, unsigned long long, bool
```
      
That is the reason why I cast to `unsigned int` first. Let me know your thoughts

shubhe25p commented:
> Could you figure out why is our UTF-32-only wcrtomb gets pulled in by this change? I.e. is there a particular function that is enabled on Windows and depends on the printf functionality?
> 
Thank you so much for the feedback, it is really helpful @vonosmas. I am not able to find this specific function, can you please help? I can add a macro to guard the wcrtomb in char_converter.h as suggested by @michaelrj-google but I am not sure.

michaelrj-google commented on libc/src/stdio/printf_core/char_converter.h:
```diff
@@ -9,6 +9,15 @@
+
+#ifndef LIBC_COPT_PRINTF_DISABLE_WIDE
+#include "src/__support/wchar/wcrtomb.h"
+#endif // LIBC_COPT_PRINTF_DISABLE_WIDE
```
ideally everything related to the wide character printing would be controlled by `LIBC_COPT_PRINTF_DISABLE_WIDE`

michaelrj-google commented on libc/src/stdio/printf_core/CMakeLists.txt:
```diff
@@ -43,6 +43,14 @@ if(NOT TARGET ${target_error_mapper})
+
+if(${LIBC_TARGET_OS} STREQUAL "windows" OR WIN32)
+  set(wchar_deps "")
+  list(APPEND wchar_compile_flags "-DLIBC_COPT_PRINTF_DISABLE_WIDE")
```
this flag should be controlled by an option in `config/config.json` similar to the other printf options. That way we can set it to false for windows platforms in the config instead of here.

vonosmas replied:
Just to confirm - with the proposed change we will not have compile options deep inside header-specific CMakeLists.txt, and have a centralized place for all compiler defines, right?

vonosmas replied:
I would like to have a single place where we construct all defines (such as `-DLIBC_COPT_PRINTF_DISABLE_WIDE`) that would be used to build ~all llvm-libc source files, instead of having a file-specific defines in subdirectories (such as under `/src/stdio/printf_core`). @michaelrj-google  - what is the recommended approach?

michaelrj-google reviewed (commented):
The code looks good overall, but there are some structural things I'd like fixed.

shubhe25p commented on libc/src/stdio/printf_core/char_converter.h:
```diff
@@ -33,7 +41,27 @@ LIBC_INLINE int convert_char(Writer<write_mode> *writer,
+
+    auto ret = internal::wcrtomb(mb_str, wc, &mbstate);
+    if (!ret.has_value()) {
+      return -1;
```
Following wctomb implementation, I think I will set libc_errno to EILSEQ and return -1. Does that sound reasonable?

shubhe25p commented on libc/test/src/stdio/sprintf_test.cpp:
```diff
@@ -3487,3 +3493,36 @@ TEST(LlvmLibcSPrintfTest, IndexModeParsing) {
+
+  // Euro sign left justified.
+  written = LIBC_NAMESPACE::sprintf(buff, "%-4lc", static_cast<wchar_t>(L'€'));
+  EXPECT_EQ(written, 6);
```
No, my wchar implementation was wrong as you pointed out, we should only print 4 character with one padding space and three byte for euro sign.

vonosmas commented on libc/src/stdio/printf_core/char_converter.h:
```diff
@@ -20,20 +29,49 @@ namespace printf_core {
-  char c = static_cast<char>(to_conv.conv_val_raw);
 
-  constexpr int STRING_LEN = 1;
+  char buffer[MB_LEN_MAX];
```
Do you need to include libc/hdr/limits_macros.h to ensure `MB_LEN_MAX` is available?

vonosmas commented on libc/test/src/stdio/sprintf_test.cpp:
```diff
@@ -3487,3 +3492,61 @@ TEST(LlvmLibcSPrintfTest, IndexModeParsing) {
+  int written;
+
+  // Euro sign is a 3-byte UTF-8 character.
+  written = LIBC_NAMESPACE::sprintf(buff, "%lc", static_cast<wchar_t>(L'€'));
```
Why do you need a static_cast<wchar_t> here? I think `L'€'` is already treated as wchar_t by compiler?

vonosmas commented on libc/test/src/stdio/sprintf_test.cpp:
```diff
@@ -3487,3 +3492,61 @@ TEST(LlvmLibcSPrintfTest, IndexModeParsing) {
+  ASSERT_STREQ_LEN(written, buff, "😀");
+
+  // WEOF test.
+  written = LIBC_NAMESPACE::sprintf(buff, "%lc", static_cast<wchar_t>(WEOF));
```
No need to cast here as well, WEOF is a wint_t, which should be passed to `sprintf`

vonosmas commented on libc/test/src/stdio/sprintf_test.cpp:
```diff
@@ -9,6 +9,11 @@
 #include "src/stdio/sprintf.h"
 
+#ifndef LIBC_COPT_PRINTF_DISABLE_WIDE
+#include "hdr/types/wchar_t.h"
```
I think you'd need `wint_t` instead of `wchar_t` - see below.

vonosmas commented on libc/src/stdio/printf_core/CMakeLists.txt:
```diff
@@ -67,6 +79,7 @@ add_header_library(
     parser.h
   DEPENDS
     .core_structs
+    libc.hdr.types.wint_t
```
It's possible you'll still need to add this dep conditionally

vonosmas commented on libc/src/stdio/printf_core/parser.h:
```diff
@@ -9,6 +9,10 @@
 #ifndef LLVM_LIBC_SRC_STDIO_PRINTF_CORE_PARSER_H
 #define LLVM_LIBC_SRC_STDIO_PRINTF_CORE_PARSER_H
 
+#ifndef LIBC_COPT_PRINTF_DISABLE_WIDE
```
Move this next to other optional includes below (e.g. the ones for `LIBC_INTERNAL_PRINTF_HAS_FIXED_POINT`)

vonosmas commented on libc/src/stdio/printf_core/char_converter.h:
```diff
@@ -20,20 +30,49 @@ namespace printf_core {
 
-  constexpr int STRING_LEN = 1;
+  char buffer[MB_LEN_MAX];
+  cpp::string_view to_write;
```
TBH I would just have `size_t write_size` here, and convert to `string_view` inside Writer invocation, since you're not actually doing anything with `string_view` before:

```
RET_IF_RESULT_NEGATIVE(writer->write({buffer, write_size}));

vonosmas reviewed (commented):
That looks mostly fine to me, thank you for addressing many comments - and see a few more comments / suggestions below.

michaelrj-google commented on libc/config/windows/config.json:
```diff
@@ -0,0 +1,8 @@
+        "doc": "Disable handling wide characters for printf and friends."
+      }
+    }
+}
```
nit: fix missing newline

michaelrj-google reviewed (commented):
Overall looks good, just minor comments

Commit:
[libc] Support %lc in printf

Add support for %lc in printf by calling internal
wcrtomb function and relevant end-to-end sprintf
test. Additionally, modified printf parser for
recognizing length modifier and added two internal
error codes. Also added a flag to disable wchar
support on windows platform.

shubhe25p commented on libc/src/stdio/printf_core/parser.h:
```diff
@@ -173,7 +177,15 @@ template <typename ArgProvider> class Parser {
       case ('c'):
-        WRITE_ARG_VAL_SIMPLEST(section.conv_val_raw, int, conv_index);
+        if (section.length_modifier == LengthModifier::l) {
+#ifndef LIBC_COPT_PRINTF_DISABLE_WIDE
```
Resolving this as @michaelrj-google didn't raise any concerns in his latest review. Feel free to open again.

shubhe25p committed:
Merge branch 'main' into br_printf_lc_support

vonosmas reviewed (approved):
Thank you for working on this!

michaelrj-google reviewed (approved):
LGTM, thanks for working on this! Do you need me to merge it for you?

shubhe25p commented:
> LGTM, thanks for working on this! Do you need me to merge it for you?

Yes please feel free to merge it and thank you so much for all the feedback you and @vonosmas provided. This would not be possible without it.

Kewen12 commented:
Hello! It seems this PR caused a link error that breaks bots. maybe need a cmake update? Could you please fix it? Thanks!

```
[4286/4884] Linking CXX executable bin/llvm-gpu-loader
FAILED: bin/llvm-gpu-loader 
: && /usr/bin/c++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-array-bounds -Wno-stringop-overread -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -Wl,-rpath-link,/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/./lib  -Wl,--gc-sections tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/llvm-gpu-loader.cpp.o tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/amdhsa.cpp.o -o bin/llvm-gpu-loader  -Wl,-rpath,"\$ORIGIN/../lib:/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/lib:/opt/rocm/lib:"  lib/libLLVMOption.so.23.0git  lib/libLLVMFrontendOffloading.so.23.0git  /opt/rocm/lib/libhsa-runtime64.so.1.14.60300  lib/libLLVMObject.so.23.0git  lib/libLLVMBinaryFormat.so.23.0git  lib/libLLVMTargetParser.so.23.0git  lib/libLLVMSupport.so.23.0git  -Wl,-rpath-link,/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/lib && :
/usr/bin/ld: tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/amdhsa.cpp.o: in function `int __llvm_libc_common_utils::printf_core::convert<(__llvm_libc_common_utils::printf_core::WriteMode)0>(__llvm_libc_common_utils::printf_core::Writer<(__llvm_libc_common_utils::printf_core::WriteMode)0>*, __llvm_libc_common_utils::printf_core::FormatSection const&)':
amdhsa.cpp:(.text._ZN24__llvm_libc_common_utils11printf_core7convertILNS0_9WriteModeE0EEEiPNS0_6WriterIXT_EEERKNS0_13FormatSectionE[_ZN24__llvm_libc_common_utils11printf_core7convertILNS0_9WriteModeE0EEEiPNS0_6WriterIXT_EEERKNS0_13FormatSectionE]+0x556): undefined reference to `__llvm_libc_common_utils::internal::wcrtomb(char*, wchar_t, __llvm_libc_common_utils::internal::mbstate*)'
```

bot: https://lab.llvm.org/buildbot/#/builders/73/builds/15649

shubhe25p commented:
> Hello! It seems this PR caused a link error that breaks bots. maybe need a cmake update? Could you please fix it? Thanks!
> 
> ```
> [4286/4884] Linking CXX executable bin/llvm-gpu-loader
> FAILED: bin/llvm-gpu-loader 
> : && /usr/bin/c++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-array-bounds -Wno-stringop-overread -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -Wl,-rpath-link,/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/./lib  -Wl,--gc-sections tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/llvm-gpu-loader.cpp.o tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/amdhsa.cpp.o -o bin/llvm-gpu-loader  -Wl,-rpath,"\$ORIGIN/../lib:/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/lib:/opt/rocm/lib:"  lib/libLLVMOption.so.23.0git  lib/libLLVMFrontendOffloading.so.23.0git  /opt/rocm/lib/libhsa-runtime64.so.1.14.60300  lib/libLLVMObject.so.23.0git  lib/libLLVMBinaryFormat.so.23.0git  lib/libLLVMTargetParser.so.23.0git  lib/libLLVMSupport.so.23.0git  -Wl,-rpath-link,/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/lib && :
> /usr/bin/ld: tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/amdhsa.cpp.o: in function `int __llvm_libc_common_utils::printf_core::convert<(__llvm_libc_common_utils::printf_core::WriteMode)0>(__llvm_libc_common_utils::printf_core::Writer<(__llvm_libc_common_utils::printf_core::WriteMode)0>*, __llvm_libc_common_utils::printf_core::FormatSection const&)':
> amdhsa.cpp:(.text._ZN24__llvm_libc_common_utils11printf_core7convertILNS0_9WriteModeE0EEEiPNS0_6WriterIXT_EEERKNS0_13FormatSectionE[_ZN24__llvm_libc_common_utils11printf_core7convertILNS0_9WriteModeE0EEEiPNS0_6WriterIXT_EEERKNS0_13FormatSectionE]+0x556): undefined reference to `__llvm_libc_common_utils::internal::wcrtomb(char*, wchar_t, __llvm_libc_common_utils::internal::mbstate*)'
> ```
> 
> bot: https://lab.llvm.org/buildbot/#/builders/73/builds/15649

Yes looking into it! Apologies I wasn't aware of the AMD openmp tests.

Kewen12 commented:
> > Hello! It seems this PR caused a link error that breaks bots. maybe need a cmake update? Could you please fix it? Thanks!
> > ```
> > [4286/4884] Linking CXX executable bin/llvm-gpu-loader
> > FAILED: bin/llvm-gpu-loader 
> > : && /usr/bin/c++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-array-bounds -Wno-stringop-overread -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -Wl,-rpath-link,/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/./lib  -Wl,--gc-sections tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/llvm-gpu-loader.cpp.o tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/amdhsa.cpp.o -o bin/llvm-gpu-loader  -Wl,-rpath,"\$ORIGIN/../lib:/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/lib:/opt/rocm/lib:"  lib/libLLVMOption.so.23.0git  lib/libLLVMFrontendOffloading.so.23.0git  /opt/rocm/lib/libhsa-runtime64.so.1.14.60300  lib/libLLVMObject.so.23.0git  lib/libLLVMBinaryFormat.so.23.0git  lib/libLLVMTargetParser.so.23.0git  lib/libLLVMSupport.so.23.0git  -Wl,-rpath-link,/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/lib && :
> > /usr/bin/ld: tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/amdhsa.cpp.o: in function `int __llvm_libc_common_utils::printf_core::convert<(__llvm_libc_common_utils::printf_core::WriteMode)0>(__llvm_libc_common_utils::printf_core::Writer<(__llvm_libc_common_utils::printf_core::WriteMode)0>*, __llvm_libc_common_utils::printf_core::FormatSection const&)':
> > amdhsa.cpp:(.text._ZN24__llvm_libc_common_utils11printf_core7convertILNS0_9WriteModeE0EEEiPNS0_6WriterIXT_EEERKNS0_13FormatSectionE[_ZN24__llvm_libc_common_utils11printf_core7convertILNS0_9WriteModeE0EEEiPNS0_6WriterIXT_EEERKNS0_13FormatSectionE]+0x556): undefined reference to `__llvm_libc_common_utils::internal::wcrtomb(char*, wchar_t, __llvm_libc_common_utils::internal::mbstate*)'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > bot: https://lab.llvm.org/buildbot/#/builders/73/builds/15649
> 
> Yes looking into it! Apologies I wasn't aware of the AMD openmp tests.

Thanks! If anything we can help with, feel feel to let us know :)

shubhe25p commented:
> > > Hello! It seems this PR caused a link error that breaks bots. maybe need a cmake update? Could you please fix it? Thanks!
> > > ```
> > > [4286/4884] Linking CXX executable bin/llvm-gpu-loader
> > > FAILED: bin/llvm-gpu-loader 
> > > : && /usr/bin/c++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-array-bounds -Wno-stringop-overread -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -Wl,-rpath-link,/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/./lib  -Wl,--gc-sections tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/llvm-gpu-loader.cpp.o tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/amdhsa.cpp.o -o bin/llvm-gpu-loader  -Wl,-rpath,"\$ORIGIN/../lib:/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/lib:/opt/rocm/lib:"  lib/libLLVMOption.so.23.0git  lib/libLLVMFrontendOffloading.so.23.0git  /opt/rocm/lib/libhsa-runtime64.so.1.14.60300  lib/libLLVMObject.so.23.0git  lib/libLLVMBinaryFormat.so.23.0git  lib/libLLVMTargetParser.so.23.0git  lib/libLLVMSupport.so.23.0git  -Wl,-rpath-link,/home/botworker/bbot/openmp-offload-libc-amdgpu-runtime/build/lib && :
> > > /usr/bin/ld: tools/llvm-gpu-loader/CMakeFiles/llvm-gpu-loader.dir/amdhsa.cpp.o: in function `int __llvm_libc_common_utils::printf_core::convert<(__llvm_libc_common_utils::printf_core::WriteMode)0>(__llvm_libc_common_utils::printf_core::Writer<(__llvm_libc_common_utils::printf_core::WriteMode)0>*, __llvm_libc_common_utils::printf_core::FormatSection const&)':
> > > amdhsa.cpp:(.text._ZN24__llvm_libc_common_utils11printf_core7convertILNS0_9WriteModeE0EEEiPNS0_6WriterIXT_EEERKNS0_13FormatSectionE[_ZN24__llvm_libc_common_utils11printf_core7convertILNS0_9WriteModeE0EEEiPNS0_6WriterIXT_EEERKNS0_13FormatSectionE]+0x556): undefined reference to `__llvm_libc_common_utils::internal::wcrtomb(char*, wchar_t, __llvm_libc_common_utils::internal::mbstate*)'
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > bot: https://lab.llvm.org/buildbot/#/builders/73/builds/15649
> > 
> > Yes looking into it! Apologies I wasn't aware of the AMD openmp tests.
> 
> Thanks! If anything we can help with, feel feel to let us know :)

Yes sure I will provide an update soon, will reach out if I am stuck. Thanks a ton!
